resilience4j-rxjava2/src/main/java/io/github/resilience4j/circuitbreaker/operator/CircuitBreakerSubscriber.java

```diff
@@ -0,0 +1,123 @@
+package io.github.resilience4j.circuitbreaker.operator;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+import io.github.resilience4j.adapter.Permit;
+import io.github.resilience4j.circuitbreaker.CircuitBreaker;
+import io.github.resilience4j.circuitbreaker.CircuitBreakerOpenException;
+import io.github.resilience4j.core.StopWatch;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A RxJava {@link Subscriber} to protect another subscriber by a CircuitBreaker.
+ *
+ * @param <T> the value type of the upstream and downstream
+ */
+final class CircuitBreakerSubscriber<T> implements Subscriber<T>, Subscription {
+    private static final Logger LOG = LoggerFactory.getLogger(CircuitBreakerSubscriber.class);
+    private final CircuitBreaker circuitBreaker;
+    private final Subscriber<? super T> childSubscriber;
+    private Subscription subscription;
+    private final AtomicBoolean cancelled = new AtomicBoolean(false);
+    private final AtomicReference<Permit> permitted = new AtomicReference<>(Permit.PENDING);
+    private StopWatch stopWatch;
+
+    CircuitBreakerSubscriber(CircuitBreaker circuitBreaker, Subscriber<? super T> childSubscriber) {
+        this.circuitBreaker = requireNonNull(circuitBreaker);
+        this.childSubscriber = requireNonNull(childSubscriber);
+    }
+
+    @Override
+    public void onSubscribe(Subscription subscription) {
+        this.subscription = subscription;
+        LOG.debug("onSubscribe");
+        if (acquireCallPermit()) {
+            childSubscriber.onSubscribe(this);
+        } else {
+            subscription.cancel();
+            childSubscriber.onSubscribe(this);
+            childSubscriber.onError(new CircuitBreakerOpenException(String.format("CircuitBreaker '%s' is open", circuitBreaker.getName())));
+        }
+    }
+
+    @Override
+    public void onNext(T event) {
+        LOG.debug("onNext: {}", event);
+        if (isInvocationPermitted()) {
+            childSubscriber.onNext(event);
+        }
+    }
+
+    @Override
+    public void onError(Throwable e) {
+        LOG.debug("onError", e);
+        markFailure(e);
+        if (isInvocationPermitted()) {
+            childSubscriber.onError(e);
+        }
+    }
+
+    @Override
+    public void onComplete() {
+        LOG.debug("onComplete");
+        markSuccess();
```

Comment:
I'm not 100% sure about this one but looked logical to me to mark the call as success when the subscription was cancelled after subscribing to it. For example some elements could have been already emitted in onNext but the call wasn't registered in CB. Unfortunately, it depends on upstream whether it honors cancellation or not. 
The same logic was applied in all other CB operators. 
What do you think about this?

Reply:
For me it looks OK, but I think we should explicitly mention this behaviour in documentation.